spec/models/course/condition/achievement_spec.rb

```diff
@@ -92,6 +92,24 @@
             user_achievement.save!
           end
         end
+
+        context 'when the user achievement is destroyed' do
+          it 'evaluate_conditional_for the affected course_user' do
+            user_achievement = create(:course_user_achievement, course_user: course_user)
+            expect(Course::Condition::Achievement).
+              to receive(:evaluate_conditional_for).with(course_user)
+            user_achievement.destroy!
+          end
+        end
+
+        context 'when the user achievement is destroyed through update attributes' do
+          it 'evaluate_conditional_for the affected course_user' do
+            user_achievement = create(:course_user_achievement, course_user: course_user)
+            expect(Course::Condition::Achievement).
+              to receive(:evaluate_conditional_for).with(course_user)
+            user_achievement.achievement.update_attributes(course_user_ids: [])
```

Comment:
@allenwq Yes, it doesn't call the callbacks of course_user_achievement unless the dependent: :destroy is added in the previous part. Controller are more likely to depends on course_user_ids to update the has_many through relationship. https://github.com/Coursemology/coursemology2/blob/master/app/controllers/course/achievement/achievements_controller.rb